engine/grid.py

```diff
@@ -174,15 +354,55 @@ def get_border_lawnmower_waypoints(self):
             if i % 2 == 0:
                 node1 = node_list[0, i]
                 node2 = node_list[rows - 1, i]
+                if node1.is_active:
+                    self.active_nodes_list.append(node1)
+                else:
+                    self.inactive_nodes_list.append(node1)
+                if node2.is_active:
+                    self.active_nodes_list.append(node2)
+                else:
+                    self.inactive_nodes_list.append(node2)
                 waypoints.append(node1)
                 waypoints.append(node2)
             elif i % 2 == 1:
                 node1 = node_list[rows - 1, i]
                 node2 = node_list[0, i]
+                if node1.is_active:
+                    self.active_nodes_list.append(node1)
+                else:
+                    self.inactive_nodes_list.append(node1)
+                if node2.is_active:
+                    self.active_nodes_list.append(node2)
+                else:
+                    self.inactive_nodes_list.append(node2)
                 waypoints.append(node1)
                 waypoints.append(node2)
         return waypoints
 
+    def get_straight_line_waypoints(self,y_start_pct):
+        """
+        Returns the robot's lawnmower border traversal path for the current grid using
+        only nodes in a straight line . Starting node is the left-most node starting at
+        the row in [Node List] that corresponds to how high the straight line should start at.
+
+        PARAMETERS:
+        ---------
+        y_start_pct: What percentage height that we want to start the straight line.(ex: 0.5: if there are 20 rows,
+                     start straight line at the 10th row up)
+        """
+        waypoints = []
+        node_list = self.nodes
+        rows = node_list.shape[0]
+        cols = node_list.shape[1]
+        selected_row = int(y_start_pct*rows)
+        for i in range(cols):
+            if node_list[selected_row][i].is_active:
```

Comment:
since node_list[selected_row][i].is_active is used in both branch cases & after, i think it would be cleaner/easier to read if it's stored into a variable before

Reply:
@Shreya1233